lib/libpciaccess/src/scanpci.man
.\" Copyright (C) 2000 The XFree86 Project, Inc.  All Rights Reserved.
.\" 
.\" Permission is hereby granted, free of charge, to any person obtaining a copy
.\" of this software and associated documentation files (the "Software"), to
.\" deal in the Software without restriction, including without limitation the
.\" rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
.\" sell copies of the Software, and to permit persons to whom the Software is
.\" furnished to do so, subject to the following conditions:
.\" 
.\" The above copyright notice and this permission notice shall be included in
.\" all copies or substantial portions of the Software.
.\" 
.\" THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
.\" IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
.\" FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
.\" XFREE86 PROJECT BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
.\" IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
.\" CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
.\" 
.\" Except as contained in this notice, the name of the XFree86 Project shall
.\" not be used in advertising or otherwise to promote the sale, use or other
.\" dealings in this Software without prior written authorization from the
.\" XFree86 Project.
.\" 
.TH SCANPCI 1 __xorgversion__
.SH NAME
scanpci - scan/probe PCI buses
.SH SYNOPSIS
.B scanpci
.RB [ \-v ]
.SH DESCRIPTION
.I Scanpci
is a utility that can be used to scan PCI buses and report information
about the configuration space settings for each PCI device.
On most platforms,
.I scanpci
can only be run by the root user.
.SH OPTIONS
.TP 8
.B \-v
Print the configuration space information for each device in a verbose
format.  Without this option, only a brief description is printed for
each device.



xserver/ChangeLog
commit d63efecc9471ac53535932b80a85b7f408f06fb9
Author: Daniel Stone <daniel@fooishbar.org>
Date:   Wed Dec 12 21:57:59 2007 +0200

    Bump to 1.4.0.90

commit 446efcc554195970cb3ddcd992f7aac617d45b1d
Author: Bartosz Fabianowski <freebsd@chillt.de>
Date:   Fri Dec 7 02:38:14 2007 +0000

    Input: Fix proximity events with valuators
    
    Initialise num_events to 1, so we always send a proximity event, and then
    optionally valuator events.  Also make sure mieq can deal with valuator
    events sent after proximity events.
    (cherry picked from commit 2dcfab37d38c0c72e9be7cc724047405c8029e88)

commit 9f4689173ef9db080592497dc2212ae79b8d6e02
Author: Daniel Stone <daniel@fooishbar.org>
Date:   Thu Dec 6 00:46:32 2007 +0000

    KDrive: Xephyr: Fix non-GLX builds
    
    Only set noGlxExtension if we're actually building GLX.

commit d37351308b255d5f9bff3438b6767c62974902da
Author: Daniel Stone <daniel@fooishbar.org>
Date:   Wed Dec 5 19:37:48 2007 +0000

    XKB: Actions: Don't run certain actions on the core keyboard
    
    Don't run VT switches, terminations, or anything, on the core keyboard: only
    run actions which affect the keyboard state.  If we get an action such as VT
    switch, just swallow the event.
    (cherry picked from commit 320abd7d1d906807448fa01ad3377daf707f46cc)

commit 27da1367c9ea143946b8b8d3dbd0f9d44c4a9039
Author: Daniel Stone <daniel@fooishbar.org>
Date:   Wed Dec 5 19:36:59 2007 +0000

    WaitForSomething: Ignore EAGAIN
    
    If select ever returns EAGAIN, don't bother complaining.
    (cherry picked from commit 85dd8efac1bc0715f03c99d261b1c5d0980623e1)

commit 259f86b13b453f3503afd3d523de32b43996d334
Author: Rich Coe <Richard.Coe@med.ge.com>
Date:   Wed Dec 5 19:36:37 2007 +0000

    OS: Connection: Keep trying select while it gets interrupted (bug #9240)
    
    If we got interrupted (EINTR or EAGAIN) during select, just try again, rather
    than shutting clients down on either of these errors.
    (cherry picked from commit b7f3618f3933a810778093fd47564a1e3bf3fde6)

commit 90649e6a39dc6caad8313b25ef869a089f81aba7
Author: Rich Coe <Richard.Coe@med.ge.com>
Date:   Wed Dec 5 19:31:07 2007 +0000

    OS: Connection: Don't shut down disappeared clients (bug #7876)
    
    If a client disappears in the middle of CheckConnections (presumably
    because its appgroup leader disappears), then don't attempt to shut it down
    a second time, when it's already vanished.
    (cherry picked from commit d8b2cad3771a09860e7be1726f67e684cf7caeec)

commit 25d26b55e74b50a2fd0632329cb0bdca017fe8e6
Author: Kanru Chen <koster@debian.org.tw>
Date:   Mon Dec 3 12:46:45 2007 +0000

    Config: HAL: Fix XKB option parsing
    
    Actually combine the XKB options into a string, rather than just repeatedly
    writing a comma.
    (cherry picked from commit da893908feb2dcf7c22420b3426ab3ac65c7ca99)

commit b037e4a5abb878ad89e7f27c2b6c23004625f6c3
Author: Peter Harris <peter.harris@hummingbird.com>
Date:   Mon Oct 29 18:05:19 2007 -0400

    Add missing swaps in panoramiXSwap.c
    (cherry picked from commit cb67a10b7f6f564e0345de19316934361ea28720)

commit 3e0993fcf38e47dd42c27a2dcb5dde7d23222ca8
Author: Daniel Stone <daniel@fooishbar.org>
Date:   Fri Nov 30 20:35:26 2007 +0200

    ProcessOtherEvent: Don't do double translation of button events
    
    We already deal with the button mapping in GetPointerEvents, so don't
    do the remapping again in ProcessOtherEvent.
    (cherry picked from commit 7ff002fe3e229330216d7f2ff16cdabe63014bcd)

commit cbf775cde7bb737ddf71fa3aa5b08c859d516084
Author: Peter Hutterer <peter@cs.unisa.edu.au>
Date:   Sat Nov 17 22:50:07 2007 +0100

    XKB: Generate correct key repeat events (bug #13114)
    
    Make sure we send the correct event for the type of device when we're
    sending key repeat events, which stops repeats being sent to incorrect
    windows.

commit 3e987ea670aadefeb3a6ad05d9a39dd7902985f9
Author: Michel Dänzer <michel@tungstengraphics.com>
Date:   Thu Oct 18 17:44:14 2007 +0200

    EXA: Don't attempt to move in pixmaps that can't be accelerated.
    
    Fixes https://bugs.freedesktop.org/show_bug.cgi?id=12815 .
    
    (Related to commit 5d74416740de883b7ef0994afea4bbd4d3901be0 on master.)

commit 75b9dc907b332d64d074083cae0c6b099960f09b
Author: Michel Dänzer <michel@tungstengraphics.com>
Date:   Thu Sep 27 13:08:41 2007 +0200

    EXA: Make sure tile offsets passed to drivers are never negative.
    
    Thanks to Björn Steinbrink for pointing out the problem on IRC.
    
    (cherry picked from commit 006f6525057970a74382132237b2131286ad147c with
     modifications.)

commit 732d586b0919e57ed836999f4117db3e776e2934
Author: Michel Dänzer <michel@tungstengraphics.com>
Date:   Thu Sep 27 13:08:40 2007 +0200

    EXA: Punt on fallback case not handled correctly in exaFillRegionTiled.
    
    Fixes http://bugs.freedesktop.org/show_bug.cgi?id=12520 .
    
    (From master commit c7d6d1f5, modified to suit.)

commit a3aed33244914b64d08630e19100c71ab81e1a81
Author: Daniel Stone <daniel@fooishbar.org>
Date:   Sat Nov 17 22:34:47 2007 +0100

    XKB: Don't ring the bell when we don't have a BellProc (bug #13246)
    (cherry picked from commit 55888552769ce6361174285b09dfb78ee22c170d)

commit f3a5d67688a0f691ef23cb44b1fdda190b5b8bef
Author: Daniel Stone <daniel@fooishbar.org>
Date:   Sun Sep 23 12:43:31 2007 +0300

    GetKeyboardEvents: Reject out-of-range keycodes (bug #12528)
    
    We can only deal with keycodes between 8 and 255, so make sure that we never
    accept anything out of this range.
    (cherry picked from commit 0e800ca4651a947ccef239e6fe7bf64aab92257c)

commit 35bf7c738a8286a382aeef38c0f035773b3ab96a
Author: Naoki Hamada <nao@tom-yam.or.jp>
Date:   Thu Oct 25 18:45:50 2007 +0300

    Input: Fix key down test (bug #12858)
    
    Fix the botched previous key_is_down test, which would give false positives.
    Also move key_autorepeats to a separate inline function.

lib/libpciaccess/src/scanpci.c
/*
 * (C) Copyright IBM Corporation 2006
 * All Rights Reserved.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a
 * copy of this software and associated documentation files (the "Software"),
 * to deal in the Software without restriction, including without limitation
 * on the rights to use, copy, modify, merge, publish, distribute, sub
 * license, and/or sell copies of the Software, and to permit persons to whom
 * the Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice (including the next
 * paragraph) shall be included in all copies or substantial portions of the
 * Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT.  IN NO EVENT SHALL
 * IBM AND/OR THEIR SUPPLIERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
 * DEALINGS IN THE SOFTWARE.
 */

#include <stdlib.h>
#include <stdio.h>
#include <err.h>

#include "pciaccess.h"


static void
print_pci_bridge( const struct pci_bridge_info * info )
{
    printf( "  Bus: primary=%02x, secondary=%02x, subordinate=%02x, "
	    "sec-latency=%u\n",
	    info->primary_bus,
	    info->secondary_bus,
	    info->subordinate_bus,
	    info->secondary_latency_timer );
    printf( "  I/O behind bridge: %08x-%08x\n",
	    info->io_base,
	    info->io_limit );
    printf( "  Memory behind bridge: %08x-%08x\n",
	    info->mem_base,
	    info->mem_limit );
    printf( "  Prefetchable memory behind bridge: %08llx-%08llx\n",
	    info->prefetch_mem_base,
	    info->prefetch_mem_limit );
}

static void
print_pci_device( struct pci_device * dev, int verbose )
{
    const char * dev_name;
    const char * vend_name;

    vend_name = pci_device_get_vendor_name( dev );
    dev_name = pci_device_get_device_name( dev );
    if ( dev_name == NULL ) {
	dev_name = "Device unknown";
    }

    printf("\npci ");
    if (dev->domain != 0)
	printf("domain 0x%04x ", dev->domain);
    printf("bus 0x%04x cardnum 0x%02x function 0x%02x:"
	   " vendor 0x%04x device 0x%04x\n",
	   dev->bus, 
	   dev->dev,
	   dev->func,
	   dev->vendor_id,
	   dev->device_id );
    if ( vend_name != NULL ) {
	printf( " %s %s\n", vend_name, dev_name );
    }
    else {
	printf( " %s\n", dev_name );
    }
    
    if ( verbose ) {
	unsigned   i;
	uint16_t  command, status;
	uint8_t   bist;
	uint8_t   header_type;
	uint8_t   latency_timer;
	uint8_t   cache_line_size;
	uint8_t   max_latency;
	uint8_t   min_grant;
	uint8_t   int_pin;


	vend_name = pci_device_get_subvendor_name( dev );
	dev_name = pci_device_get_subdevice_name( dev );
	if ( dev_name == NULL ) {
	    dev_name = "Card unknown";
	}

	printf( " CardVendor 0x%04x card 0x%04x (",
		dev->subvendor_id,
		dev->subdevice_id );
	if ( vend_name != NULL ) {
	    printf( "%s, %s)\n", vend_name, dev_name );
	}
	else {
	    printf( "%s)\n", dev_name );
	}

	pci_device_cfg_read_u16( dev, & command, 4 );
	pci_device_cfg_read_u16( dev, & status,  6 );
	printf( "  STATUS    0x%04x  COMMAND 0x%04x\n", 
		status,
		command );
	printf( "  CLASS     0x%02x 0x%02x 0x%02x  REVISION 0x%02x\n",
		(dev->device_class >> 16) & 0x0ff,
		(dev->device_class >>  8) & 0x0ff,
		(dev->device_class >>  0) & 0x0ff,
		dev->revision );

	pci_device_cfg_read_u8( dev, & cache_line_size, 12 );
	pci_device_cfg_read_u8( dev, & latency_timer, 13 );
	pci_device_cfg_read_u8( dev, & header_type, 14 );
	pci_device_cfg_read_u8( dev, & bist, 15 );

	printf( "  BIST      0x%02x  HEADER 0x%02x  LATENCY 0x%02x  CACHE 0x%02x\n",
		bist,
		header_type,
		latency_timer,
		cache_line_size );
	
	pci_device_probe( dev );
	for ( i = 0 ; i < 6 ; i++ ) {
	    if ( dev->regions[i].base_addr != 0 ) {
		printf( "  BASE%u     0x%08x SIZE %d  %s",
			i,
			(intptr_t) dev->regions[i].base_addr,
			(size_t) dev->regions[i].size,
			(dev->regions[i].is_IO) ? "I/O" : "MEM" );

		if ( ! dev->regions[i].is_IO ) {
		    if ( dev->regions[i].is_prefetchable ) {
			printf( " PREFETCHABLE" );
		    }
		}
		
		printf( "\n" );
	    }
	}

	if ( dev->rom_size ) {
	    printf( "  BASEROM   0x%08x  addr 0x%08x\n",
		    0, 0 );
	}

	pci_device_cfg_read_u8( dev, & int_pin, 61 );
	pci_device_cfg_read_u8( dev, & min_grant, 62 );
	pci_device_cfg_read_u8( dev, & max_latency, 63 );

	printf( "  MAX_LAT   0x%02x  MIN_GNT 0x%02x  INT_PIN 0x%02x  INT_LINE 0x%02x\n",
		max_latency,
		min_grant,
		int_pin,
		dev->irq );

	if ( (dev->device_class >> 16) == 0x06 ) {
	    const void * info;

	    if ( (info = pci_device_get_bridge_info(dev)) != NULL ) {
		print_pci_bridge( (const struct pci_bridge_info *) info );
	    }
	    else if ( (info = pci_device_get_pcmcia_bridge_info(dev)) != NULL ) {
		/* Nothing yet. */
	    }
	}
    }
}


int main( int argc, char ** argv )
{
    struct pci_device_iterator * iter;
    struct pci_device * dev;
    int ret;
    int verbose = 0;
    int c;
    int errors = 0;

    while ((c = getopt(argc, argv, "v")) != -1) {
	switch (c) {
	case 'v':
	    verbose = 1;
	    break;
	case '?':
	    errors++;
	}
    }
    if (errors != 0) {
	fprintf(stderr, "usage: %s [-v]\n", argv[0]);
	exit(2);
    }

    ret = pci_system_init();
    if (ret != 0)
	err(1, "Couldn't initialize PCI system");

    iter = pci_slot_match_iterator_create( NULL );

    while ( (dev = pci_device_next( iter )) != NULL ) {
	print_pci_device( dev, verbose );
    }

    pci_system_cleanup();
    return 0;
}
